clc/modules/object-storage-common/src/main/java/com/eucalyptus/objectstorage/exceptions/s3/S3Exceptions.groovy

```diff
@@ -197,6 +219,73 @@ class BucketNotEmptyException extends S3Exception {
   }
 }
 
+// On a CORS PUT of a new CORS configuration, if a bad HTTP verb (e.g. "YUCK")
+// or an unsupported HTTP verb (e.g. "OPTIONS") is provided,
+// to match AWS's response, return 400 Bad Request with this error code
+// and message.
+class CorsConfigUnsupportedMethodException extends S3Exception {
+  def CorsConfigUnsupportedMethodException(String method) {
+    super(S3ErrorCodeStrings.InvalidRequest,
+    "Found unsupported HTTP method in CORS config. Unsupported method is " + method,
+    HttpResponseStatus.BAD_REQUEST);
+  }
+}
+
+// On a CORS preflight OPTIONS request, if no HTTP verb (method) is provided,
+// or if a bad HTTP verb (e.g. "YUCK") or an unsupported HTTP verb
+// (e.g. "OPTIONS") is provided,
+// to match AWS's response, return 400 Bad Request with this error code
+// and message.
+class CorsPreflightInvalidMethodException extends S3Exception {
+  def CorsPreflightInvalidMethodException(String method) {
+    super(S3ErrorCodeStrings.BadRequest,
+    "Invalid Access-Control-Request-Method: " + method,
+    HttpResponseStatus.BAD_REQUEST);
+  }
+}
+
+// On a CORS preflight OPTIONS request, when no CORS configuration exists,
+// to match AWS's response, return 403 Forbidden with this error code
+// and message.
+class CorsPreflightNoConfigException extends S3ExtendedException {
+  def CorsPreflightNoConfigException(String requestMethod, String resourceType) {
+    super(S3ErrorCodeStrings.AccessForbidden,
+    "CORSResponse: CORS is not enabled for this bucket.",
+    HttpResponseStatus.FORBIDDEN);
+    this.setResourceType(resourceType);
+    this.setRequestMethod(requestMethod);
+  }
+}
+
+// On a CORS preflight OPTIONS request, if no origin is provided,
+// to match AWS's response, return 400 Bad Request with this error code
+// and message.
+class CorsPreflightNoOriginException extends S3Exception {
+  def CorsPreflightNoOriginException() {
+    super(S3ErrorCodeStrings.BadRequest,
+    "Insufficient information. Origin request header needed.",
+    HttpResponseStatus.BAD_REQUEST);
+  }
+}
+
+// On a CORS preflight OPTIONS request, if no CORS rule matches the
+// requested origin and HTTP verb (method),
+// to match AWS's response, return 403 Forbidden with this error code
+// and message.
+class CorsPreflightNotAllowedException extends S3ExtendedException {
+  def CorsPreflightNotAllowedException(String requestMethod, String resourceType) {
+    super(S3ErrorCodeStrings.AccessForbidden,
+    "CORSResponse: This CORS request is not allowed. " +
+    "This is usually because the evaluation of Origin, request method / " +
+    "Access-Control-Request-Method or Access-Control-Request-Headers are " +
+    "not whitelisted by the resource's CORS spec.",
+    HttpResponseStatus.FORBIDDEN);
+    this.setResourceType(resourceType);
+    this.setRequestMethod(requestMethod);
+  }
+}
+
+
```

Comment:
`S3Exceptions.groovy` is exclusively for S3 defined list of [error responses]( http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#ErrorCodeList). Eucalyptus specific exceptions used internally for convenience should be defined in another class, probably under `com.eucalyptus.objectstorage.exceptions` package